### PR TITLE
experimental externalisation on the router service via configuration

### PR DESCRIPTION
**Pull request by garypen**
fixes: #1916 

Simplified externalisation configuration:

```
plugins:
  experimental.external:
    url: http://127.0.0.1:8081
    timeout: 2s
    stages:
      router:
        request:
          headers: true
          context: true
          body: true
          sdl: true
        response:
          headers: true
```
The biggest pro is that it's just so simple it's very easy to configure. The biggest con is lack of control: e.g. no control over header propagation. The only supported stage at this moment is "router".

Todo:
 - [x] response externalisation
 - [x] agree the simplified configuration format
 - [x] tests
 - [ ] More tests based on router_service test harness enhancements/changes
 - [x] Add a timeout to the external call
 - [x] documentation
 - [x] performance
 - [x] license check
 - [x] add an example

---

configuration: discussed in 9/12 huddle. No substantial objections to this as experimental

performance

Simple load testing with a minimal co-processor (which just echoed responses) showed that:
 - baseline (no co-processor): 24k tps
 - externalising request: 17k tps
 - externalising request and response: 12k tps

### CODE REVIEW

**Comment by o0Ignition0o** [comment]
This looks great! I m going to take this one over, and move the url to the stages section

**Comment by o0Ignition0o** [comment]
nevermind, i think ppl will be able to override it in stages if needs be.

**Comment by garypen** [comment]
Hi @o0Ignition0o . Thanks for taking this over the line. One thing I did think about over the weekend that might need to change is that we should not put break returns into a GrapQL error return, but simply return the bytes "as is". What do you think?

File: docs/source/configuration/external.mdx [hunk 1]
```diff
@@ -0,0 +1,348 @@
...
...tock: Boolean @join__field(graph: INVENTORY)\n  name: String @join__field(graph: PRODUCTS)\n  price: Int @join__field(graph: PRODUCTS)\n  reviews: [Review] @join__field(graph: REVIEWS)\n  reviewsForAuthor(authorID: ID!): [Review] @join__field(graph: REVIEWS)\n  shippingEstimate: Int @join__field(graph: INVENTORY, requires: \"price weight\")\n  upc: String! @join__field(graph: PRODUCTS)\n  weight: Int @join__field(graph: PRODUCTS)\n}\n\ntype Query {\n  me: User @join__field(graph: ACCOUNTS)\n  topProducts(first: Int = 5): [Product] @join__field(graph: PRODUCTS)\n}\n\ntype Review\n  @join__owner(graph: REVIEWS)\n  @join__type(graph: REVIEWS, key: \"id\")\n{\n  author: User @join__field(graph: REVIEWS, provides: \"username\")\n  body: String @join__field(graph: REVIEWS)\n  id: ID! @join__field(graph: REVIEWS)\n  product: Product @join__field(graph: REVIEWS)\n}\n\ntype User\n  @join__owner(graph: ACCOUNTS)\n  @join__type(graph: ACCOUNTS, key: \"id\")\n  @join__type(graph: REVIEWS, key: \"id\")\n{\n  id: ID! @join__field(graph: ACCOUNTS)\n  name: String @join__field(graph: ACCOUNTS)\n  reviews: [Review] @join__field(graph: REVIEWS)\n  username: String @join__field(graph: ACCOUNTS)\n}\n"
```
Comment (o0Ignition0o) [inline docs/source/configuration/external.mdx:297]: what would happen if the sdl changes ? ^^

**Review by o0Ignition0o — APPROVED** [review]
Solid work!

LGTM, there's a few rough edges IMO but this is really great work, and will alow for experimentation.

**Comment by o0Ignition0o** [comment]
> One thing I did think about over the weekend that might need to change is that we should not put break returns into a GrapQL error return, but simply return the bytes "as is".

This is IMO part of a bigger conversation around the way we deal with errors in general, let's discuss this in a followup.

**Pull request merged**